### PR TITLE
run-make: explaing why fmt-write-bloat is ignore-windows

### PR DESCRIPTION
**Pull request by ChrisDenton**
The trouble here is that libc doesn't exist on Windows. Well it kinda does but it isn't called that so we substitute a name that works. Ideally finding necessary libs for the platform would be done at a higher level but until then this should work.

try-job: x86_64-msvc
try-job: x86_64-mingw
try-job: i686-msvc
try-job: i686-mingw

### CODE REVIEW

**Review by jieyouxu — APPROVED** [review]
Thanks, r=me if try job comes back green.

**Comment by ChrisDenton** [comment]
> warning LNK4098: defaultlib 'msvcrt' conflicts with use of other libs

Do we link dynamically on i686 but statically on x86_64? That's going to make things slightly more complicated.

File: tests/run-make/fmt-write-bloat/main.rs [hunk 1]
```diff
@@ -5,7 +5,9 @@
 use core::fmt;
 use core::fmt::Write;
 
-#[link(name = "c")]
+#[cfg_attr(not(target_env = "msvc"), link(name = "c"))]
+#[cfg_attr(all(target_env = "msvc", target_feature = "crt-static"), link(name = "libcmt"))]
+#[cfg_attr(all(target_env = "msvc", not(target_feature = "crt-static")), link(name = "msvcrt"))]
```
Comment (jieyouxu) [inline tests/run-make/fmt-write-bloat/main.rs:10]: Remark: this is not obvious, at all lol

**Review by jieyouxu — APPROVED** [review]
Pog!

**Comment by jieyouxu** [comment]
Feel free to r=me, I haven't r+'d just in case you wanted to make any changes.

**Comment by jieyouxu** [comment]
Looks good as is, so
@bors r+ rollup

File: tests/run-make/fmt-write-bloat/main.rs (hunk 1 above)
Comment (ChrisDenton) [reply tests/run-make/fmt-write-bloat/main.rs:10]: Aye. I do think this logic should ideally either be in run-make-support or better yet compiletest could have something like a `NO_STD_EXTRA_ARGS` variable that any test can use. Or at least I think that's better than individual tests needing to figure it out. But I'm not entirely sure how best to approach that.

Comment (jieyouxu) [reply tests/run-make/fmt-write-bloat/main.rs:10]: Good point. I opened #128821 so this doesn't get lost in a review comment once the PR is merged.

**Comment by compiler-errors** [comment]
https://github.com/rust-lang/rust/pull/128819#issuecomment-2275968716

@bors r-

**Comment by jieyouxu** [comment]
Wait how does this pass on msvc jobs then?

```
warning: the feature `lang_items` is internal to the compiler or standard library
```

**Comment by ChrisDenton** [comment]
> warning: the feature `lang_items` is internal to the compiler or standard library

It's just a warning, right? The thing that confuses me is how it passed the mingw try-job?

**Comment by ChrisDenton** [comment]
Oh wait I forgot about mingw originally, lol.

**Comment by jieyouxu** [comment]
> Oh wait I forgot about mingw originally, lol.

Yeah I just updated the try-jobs lol

**Comment by ChrisDenton** [comment]
Ok, I've been assuming that linking these libs is necessary. But let's see if they actually are needed.

@bors try

**Comment by ChrisDenton** [comment]
```
Symbol { name: "__imp__OutputDebugStringA@4", address: 4264468, size: 0, kind: Data, section: Section(SectionIndex(6)), scope: Linkage, weak: false, flags: None } contains Debug
thread 'main' panicked at C:\a\rust\rust\tests\run-make\fmt-write-bloat\rmake.rs:33:5:
assertion failed: !any_symbol_contains(&bin_name("main"), &panic_syms)
```

Ah, a proper test failure. So the good news is that we don't need all the `#[link]` shenanigans. The other news is that I've found the reason why we were ignoring `__imp_` symbols. Apparently it is because of i686-mingw (but not any other target).

**Comment by ChrisDenton** [comment]
Ok, this is not going to be as easy as I thought. Maybe I should try compiling it as a staticlib. However, I'd need to make sure it's still testing everything correctly.

**Comment by ChrisDenton** [comment]
@jieyouxu ok, I'm going to admit defeat here. At least for the time being. I have updated the test to explain why this test in particular is `ignore-windows`. I've also kept the changes to not link libc on Windows, even though Windows is ignore, in case it helps someone else who looks at this in the future.

Anyway, the tl;dr version of the comment is that to properly run this test on Windows MSVC we need to parse the debug information from a `.pdb` file. On Windows MINGW we (currently) can't stop a load of random runtime stuff being included so on that platform at least we'd need to be more specific than e.g. "Debug" when looking for symbols.

**Comment by jieyouxu** [comment]
Thank you for trying! Knowing more specifically why it doesn't work is very useful!

@bors r+ rollup

**Pull request merged**